include/realizations/catchment/Formulation_Manager.hpp

```diff
@@ -245,6 +245,18 @@ namespace realization {
                     return "";
             }
 
+   #ifdef NETCDF_ACTIVE
+            void close_nc_file()
+            {
+                for (std::vector<std::shared_ptr<Catchment_Formulation>>::iterator it = missing_formulation_vec.begin(); it != missing_formulation_vec.end(); ++it)
+                {
+                    if (it == missing_formulation_vec.begin()) {
+                        (*it)->get_forcing()->get_nc_file()->close();
```

Comment:
This will blow up if CSV forcing is used, won't it? And CSV forcing can be used when NetCDF support is compiled in, so the compile-time guard above doesn't prevent this.  In fact, a simulation can include some formulations that use NetCDF forcing and some that use CSV... and we're not done adding more.

Reply:
Also... have you tried this with a realization with more than one formulation? If you have a `global` and also another catchment-specific formulation that uses the same NetCDF forcing file--which is a valid configuration--I am pretty sure this will try to close the file _twice_, which may crash. (You could have any number of catchment formulations using the same NetCDF file, so you could try to close the same one a large number of times.)

Reply:
For the first question, I tested the code yesterday without the guards, that seemed to work ok locally with both netcdf and csv files. But it seemed to not passing some of the tests. So I tried to put some guards on.

For the second question, I have not. I assume we might design a test to see if all netCDF files are closed or somehow to make each pointer specific for an instance and then close each one separately.

Reply:
That surprises me... if a formulation uses a CSV forcing, then `GenericDataProvider::get_nc_file()` is going to return 0/nullptr and then `->close()` will be called on a nullptr... if that does not crash, then the NetCDF library authors were very clever in preventing it. It should crash or throw an exception.

Reply:
I have put in a check now so the close operation only performed if the ptr is not a nullptr.

Reply:
With the new commit, the code works with multiple formulations.